### ISSUE TITLE
Implement GPU Acceleration for LLM Inference and Embeddings

### ISSUE THREAD

**Issue by Excelsior2026** [eleanor-project/V8#25]
## Overview

Implement GPU acceleration across ELEANOR V8 to significantly improve performance for LLM inference, embedding computations, and critic evaluations. This work can proceed in parallel with other production readiness improvements (#16-20).

## Performance Targets

- **Embedding Operations**: 10-50x speedup
- **Critic Execution**: 3-5x speedup with GPU batching
- **Model Inference**: 2-10x speedup (model-dependent)
- **Precedent Retrieval**: 20-100x speedup for similarity search
- **Overall Throughput**: 5-15x improvement with full GPU utilization

### New Components

```
engine/gpu/
├── __init__.py              # GPU module initialization
├── manager.py               # GPUManager: device detection, allocation
├── async_ops.py             # AsyncGPUExecutor: async GPU operations
├── embeddings.py            # GPUEmbeddingCache: GPU-accelerated embeddings
├── parallelization.py       # Multi-GPU model parallelization
├── memory_optimization.py   # Memory-efficient inference (quantization, FlashAttention)
└── monitoring.py            # GPU metrics and health checks

engine/critics/
└── gpu_batch.py             # GPUBatchCriticExecutor: batched critic execution

config/
└── gpu.yaml                 # GPU configuration
```

### Integration Points

**Minimal Overlap with Other Work:**
- ✅ Independent new directory: `engine/gpu/`
- ✅ New configuration file: `config/gpu.yaml`
- ⚠️ Minor modification: `engine/engine.py` (initialization only)
- ⚠️ Optional enhancement: `engine/precedent/` (embedding operations)
- ✅ Compatible with: Caching (#16), Observability (#17), Resilience (#18)

#### 1.1 GPU Manager

**File**: `engine/gpu/manager.py`

```python
import torch
from typing import Optional, List, Dict, Any
import logging

logger = logging.getLogger(__name__)

class GPUManager:
    """
    Manage GPU resources and device allocation.
    
    Features:
    - Auto-detect CUDA, MPS (Apple Silicon), or CPU
    - Multi-GPU device management
    - Memory statistics and monitoring
    - Device health checks
    """
    
    def __init__(self, preferred_devices: Optional[List[int]] = None):
        self.device = self._detect_device()
        self.devices_available = torch.cuda.device_count() if torch.cuda.is_available() else 0
        self.preferred_devices = preferred_devices or list(range(self.devices_available))
        
        logger.info(
            "gpu_manager_initialized",
            device=str(self.device),
            cuda_available=torch.cuda.is_available(),
            device_count=self.devices_available,
            cuda_version=torch.version.cuda if torch.cuda.is_available() else None
        )
    
    def _detect_device(self) -> torch.device:
        """Auto-detect best available device."""
        if torch.cuda.is_available():
            return torch.device("cuda")
        elif hasattr(torch.backends, 'mps') and torch.backends.mps.is_available():
            return torch.device("mps")  # Apple Silicon
        return torch.device("cpu")
    
    def get_device(self, preferred_gpu: Optional[int] = None) -> torch.device:
        """Get specific GPU device or default."""
        if preferred_gpu is not None and self.devices_available > preferred_gpu:
            return torch.device(f"cuda:{preferred_gpu}")
        return self.device
    
    def memory_stats(self, device_id: int = 0) -> Dict[str, Any]:
        """Get GPU memory statistics."""
        if not torch.cuda.is_available() or device_id >= self.devices_available:
            return {"available": False}
        
        return {
            "device_id": device_id,
            "allocated_mb": torch.cuda.memory_allocated(device_id) / 1024**2,
            "reserved_mb": torch.cuda.memory_reserved(device_id) / 1024**2,
            "max_allocated_mb": torch.cuda.max_memory_allocated(device_id) / 1024**2,
            "total_mb": torch.cuda.get_device_properties(device_id).total_memory / 1024**2,
            "utilization_pct": (
                torch.cuda.memory_allocated(device_id) / 
                torch.cuda.get_device_properties(device_id).total_memory * 100
            )
        }
    
    def health_check(self) -> Dict[str, Any]:
        """Check GPU health status."""
        if not torch.cuda.is_available():
            return {"healthy": True, "mode": "cpu"}
        
        health = {"healthy": True, "devices": []}
        
        for device_id in range(self.devices_available):
            try:
                stats = self.memory_stats(device_id)
                device_health = {
                    "device_id": device_id,
                    "healthy": stats["utilization_pct"] < 95,  # Flag if >95% full
                    "memory_stats": stats
                }
                health["devices"].append(device_health)
            except Exception as e:
                logger.error(f"GPU {device_id} health check failed: {e}")
                health["healthy"] = False
        
        return health
```

**Tasks:**
- [ ] Implement GPUManager class
- [ ] Add device detection (CUDA, MPS, CPU fallback)
- [ ] Memory statistics tracking
- [ ] Health check system
- [ ] Unit tests for all devices types

#### 1.2 Async GPU Operations

**File**: `engine/gpu/async_ops.py`

```python
import asyncio
import torch
from typing import Callable, Any, List
import logging

logger = logging.getLogger(__name__)

class AsyncGPUExecutor:
    """
    Coordinate async GPU operations with CPU tasks.
    
    Features:
    - Multiple CUDA streams for parallel operations
    - Async/await integration
    - Non-blocking GPU execution
    - Stream synchronization management
    """
    
    def __init__(self, device: torch.device, num_streams: int = 4):
        self.device = device
        
        if device.type == "cuda":
            self.streams = [torch.cuda.Stream(device) for _ in range(num_streams)]
        else:
            self.streams = [None] * num_streams  # No streams for CPU/MPS
        
        self.current_stream_idx = 0
        
        logger.info(
            "async_gpu_executor_initialized",
            device=str(device),
            num_streams=num_streams if device.type == "cuda" else 0
        )
    
    def get_stream(self) -> Any:
        """Get next available CUDA stream (round-robin)."""
        if self.device.type != "cuda":
            return None
        
        stream = self.streams[self.current_stream_idx]
        self.current_stream_idx = (self.current_stream_idx + 1) % len(self.streams)
        return stream
    
    async def execute_async(self, operation: Callable, *args, **kwargs) -> Any:
        """
        Execute GPU operation asynchronously.
        
        GPU operations are non-blocking by default, but we need to
        synchronize when returning results to CPU.
        """
        stream = self.get_stream()
        
        if stream is not None:
            with torch.cuda.stream(stream):
                result = operation(*args, **kwargs)
                
                # Synchronize stream in executor to avoid blocking event loop
                await asyncio.get_event_loop().run_in_executor(
                    None, stream.synchronize
                )
        else:
            # CPU or MPS: no stream management needed
            result = operation(*args, **kwargs)
        
        return result
    
    async def batch_execute(self, operations: List[tuple]) -> List[Any]:
        """
        Execute multiple GPU operations in parallel using different streams.
        
        Args:
            operations: List of (callable, args, kwargs) tuples
        
        Returns:
            List of results in same order
        """
        tasks = []
        
        for op, args, kwargs in operations:
            task = self.execute_async(op, *args, **kwargs)
            tasks.append(task)
        
        return await asyncio.gather(*tasks)
```

**Tasks:**
- [ ] Implement AsyncGPUExecutor
- [ ] CUDA stream management
- [ ] Async/await integration
- [ ] Batch execution support
- [ ] Integration tests with real GPU operations

#### 1.3 Configuration

**File**: `config/gpu.yaml`

```yaml
# ELEANOR V8 - GPU Acceleration Configuration

gpu:
  # Enable GPU acceleration
  enabled: true
  
  # Device preference order (first available is used)
  device_preference: ["cuda", "mps", "cpu"]
  
  # Multi-GPU settings
  multi_gpu:
    enabled: false
    device_ids: [0, 1, 2, 3]  # Which GPUs to use
    strategy: "data_parallel"  # data_parallel, model_parallel, or pipeline_parallel
  
  # Memory optimization
  memory:
    # Mixed precision training (FP16/BF16)
    mixed_precision: true
    precision: "fp16"  # fp16, bf16, or fp32
    
    # Model quantization for reduced memory
    quantization:
      enabled: false
      bits: 8  # 4, 8, or 16
      method: "dynamic"  # dynamic or static
    
    # Memory limits
    max_memory_per_gpu: "24GB"
    offload_to_cpu: false
    kv_cache_size: "4GB"
  
  # Batch processing
  batching:
    enabled: true
    default_batch_size: 8
    max_batch_size: 32
    dynamic_batching: true  # Adjust batch size based on memory
  
  # Async operations
  async:
    num_streams: 4  # Number of CUDA streams for parallel ops
    prefetch: true  # Prefetch next batch while processing current
  
  # Monitoring
  monitoring:
    log_memory_stats: true
    memory_check_interval: 60  # seconds
    enable_profiling: false  # Detailed performance profiling

# LLM Backend GPU Configuration
ollama:
  gpu_layers: -1  # -1 = all layers on GPU, 0 = CPU only, N = N layers on GPU
  num_gpu: 1  # Number of GPUs to use per model
  gpu_memory_utilization: 0.9  # Max fraction of GPU memory to use

# Embedding model GPU settings  
embeddings:
  device: "cuda"  # cuda, mps, or cpu
  batch_size: 32
  cache_on_gpu: true  # Keep embedding cache on GPU
  max_cache_size_gb: 2
```

**Tasks:**
- [ ] Create GPU configuration file
- [ ] Document all options
- [ ] Add validation schema
- [ ] Environment variable overrides

#### 2.1 GPU Embedding Cache

**File**: `engine/gpu/embeddings.py`

**Features:**
- Compute embeddings on GPU
- Cache embeddings in GPU memory
- Vectorized similarity computation
- Batch embedding generation

**Performance Target**: 10-50x speedup for embedding operations

**Tasks:**
- [ ] Implement GPUEmbeddingCache
- [ ] GPU tensor-based similarity computation
- [ ] Batch embedding generation
- [ ] Integration with precedent retrieval
- [ ] Benchmarks vs CPU embeddings

#### 3.1 GPU Batch Critic Executor

**File**: `engine/critics/gpu_batch.py`

**Features:**
- Execute critics with GPU batching
- Reduce GPU kernel launch overhead
- Mixed precision inference
- Parallel critic execution across multiple GPU streams

**Performance Target**: 3-5x speedup for critic evaluations

**Tasks:**
- [ ] Implement GPUBatchCriticExecutor
- [ ] Batch preparation and tokenization
- [ ] Mixed precision support
- [ ] Integration with existing critic protocol
- [ ] Performance benchmarks

#### 4.1 Model Parallelization

**File**: `engine/gpu/parallelization.py`

**Features:**
- Data parallelism: Different inputs on different GPUs
- Model parallelism: Split large models across GPUs
- Pipeline parallelism: Different stages on different GPUs

**Tasks:**
- [ ] Implement ModelParallelRouter
- [ ] Data parallel execution
- [ ] Model parallel support for large models
- [ ] Load balancing across GPUs
- [ ] Multi-GPU benchmarks

#### 5.1 Advanced Memory Techniques

**File**: `engine/gpu/memory_optimization.py`

**Features:**
- FlashAttention integration
- 4-bit/8-bit quantization
- KV cache optimization
- Gradient checkpointing (if needed for fine-tuning)

**Tasks:**
- [ ] FlashAttention integration
- [ ] Quantization support (bitsandbytes)
- [ ] KV cache management
- [ ] Memory profiling tools
- [ ] OOM prevention strategies

#### 6.1 Engine Initialization

**File**: `engine/engine.py` (modifications)

```python
class EleanorEngineV8:
    def __init__(
        self,
        *,
        config: Optional[EngineConfig] = None,
        gpu_config: Optional[GPUConfig] = None,  # NEW
        # ... existing parameters
    ):
        self.config = config or EngineConfig()
        
        # GPU initialization
        if gpu_config and gpu_config.enabled:
            from engine.gpu import GPUManager, AsyncGPUExecutor, GPUEmbeddingCache
            
            self.gpu_manager = GPUManager(
                preferred_devices=gpu_config.device_ids
            )
            self.gpu_executor = AsyncGPUExecutor(
                device=self.gpu_manager.device,
                num_streams=gpu_config.num_streams
            )
            
            if gpu_config.enable_embedding_cache:
                self.embedding_cache_gpu = GPUEmbeddingCache(
                    device=self.gpu_manager.device,
                    max_cache_size_gb=gpu_config.embedding_cache_size_gb
                )
            
            logger.info(
                "gpu_acceleration_enabled",
                devices=self.gpu_manager.devices_available,
                device=str(self.gpu_manager.device)
            )
        else:
            self.gpu_manager = None
            self.gpu_executor = None
            self.embedding_cache_gpu = None
```

**Tasks:**
- [ ] Add GPU configuration parameter
- [ ] Initialize GPU components conditionally
- [ ] Backward compatibility (CPU fallback)
- [ ] Integration tests
- [ ] Update factory functions

#### 7.1 GPU Metrics Integration

**File**: `engine/gpu/monitoring.py`

**Features:**
- GPU memory usage metrics
- GPU utilization tracking
- Temperature monitoring
- Integration with observability framework (#17)

**Tasks:**
- [ ] GPU metrics collection
- [ ] Integration with structured logging
- [ ] Health check endpoints
- [ ] Prometheus metrics export
- [ ] Grafana dashboard template

### Required

```txt
# requirements-gpu.txt

# Core GPU support
torch>=2.1.0  # CPU version as base
torch[cuda]>=2.1.0+cu118  # CUDA 11.8 (optional)
torchvision>=0.16.0

# GPU utilities
nvidia-ml-py>=12.535.0  # GPU monitoring
gputil>=1.4.0
```

### Optional (Performance)

```txt
# requirements-gpu-optimized.txt

# Memory optimization
flash-attn>=2.5.0  # FlashAttention
bitsandbytes>=0.41.0  # Quantization
triton>=2.1.0  # CUDA kernel optimization

# Multi-GPU
accelerate>=0.25.0  # Hugging Face multi-GPU
deepspeed>=0.12.0  # Advanced parallelization

# High-performance inference
vllm>=0.2.7  # Optional: vLLM for production
tensorrt>=8.6.0  # Optional: NVIDIA TensorRT
```

### Unit Tests
- [ ] GPU manager device detection
- [ ] Memory statistics accuracy
- [ ] Async GPU operations
- [ ] Embedding cache correctness
- [ ] Batch critic execution

### Integration Tests
- [ ] Full pipeline with GPU acceleration
- [ ] CPU/GPU result consistency
- [ ] Multi-GPU execution
- [ ] Memory overflow handling
- [ ] Graceful degradation to CPU

### Performance Benchmarks
- [ ] Embedding generation speed
- [ ] Critic execution throughput
- [ ] End-to-end pipeline latency
- [ ] Memory usage profiles
- [ ] Multi-GPU scaling

### Compatibility Tests
- [ ] CUDA 11.x compatibility
- [ ] CUDA 12.x compatibility
- [ ] Apple MPS support
- [ ] CPU-only fallback
- [ ] Mixed GPU/CPU execution

## Documentation

- [ ] `docs/GPU_ACCELERATION.md` - Usage guide
- [ ] `docs/GPU_ARCHITECTURE.md` - Technical architecture
- [ ] `docs/GPU_BENCHMARKS.md` - Performance results
- [ ] `docs/GPU_TROUBLESHOOTING.md` - Common issues
- [ ] Example: `examples/gpu_example.py`
- [ ] Example: `examples/multi_gpu_example.py`

### Conflict Risk Matrix

| Other Work | File Overlap | Risk Level | Mitigation |
|------------|--------------|------------|------------|
| Caching (#16) | None | 🟢 None | GPU can use cache layer |
| Observability (#17) | None | 🟢 None | GPU integrates with logging/tracing |
| Resilience (#18) | None | 🟢 None | Circuit breakers useful for GPU OOM |
| Config Management (#11) | `engine/engine.py` | 🟡 Low | Coordinate initialization |
| Security (#13, #20) | None | 🟢 None | Independent |
| Async Resources (#19) | `engine/engine.py` | 🟡 Low | Compatible patterns |

### Branch Strategy

**Recommended**: Branch from `feature/caching-observability-resilience` (PR #23)

**Rationale**:
- GPU operations benefit from structured logging immediately
- GPU memory stats can use observability framework
- Circuit breakers useful for GPU OOM scenarios
- Cache layer useful for GPU tensors

**Alternative**: Branch from `main` for complete independence

### Communication

- Update this issue with progress
- Tag conflicts in PR descriptions
- Coordinate `engine/engine.py` changes
- Share benchmarks in comments

## Acceptance Criteria

- [ ] GPU acceleration works on NVIDIA GPUs (CUDA 11.x/12.x)
- [ ] Apple Silicon support (MPS backend)
- [ ] Graceful fallback to CPU when GPU unavailable
- [ ] 5x+ overall throughput improvement with GPU
- [ ] Memory usage stays within configured limits
- [ ] No regression in CPU-only mode
- [ ] All tests pass on both GPU and CPU
- [ ] Documentation complete
- [ ] Benchmarks demonstrate speedup
- [ ] Configuration validation works
- [ ] Health checks report GPU status

### Performance
- ✅ >5x throughput improvement
- ✅ >10x embedding speedup
- ✅ >20x precedent retrieval speedup
- ✅ <100ms P99 latency for single request

### Reliability
- ✅ 99.9% uptime with GPU acceleration
- ✅ Zero OOM crashes in production
- ✅ Graceful degradation when GPU overloaded
- ✅ CPU fallback works seamlessly

### Adoption
- ✅ Documentation enables self-service setup
- ✅ Configuration validation prevents errors
- ✅ Monitoring shows GPU utilization

## References

- PyTorch GPU Documentation: https://pytorch.org/docs/stable/cuda.html
- CUDA Programming Guide: https://docs.nvidia.com/cuda/
- FlashAttention Paper: https://arxiv.org/abs/2205.14135
- Multi-GPU Best Practices: https://pytorch.org/tutorials/intermediate/ddp_tutorial.html

## Related Issues

- Works with: #16 (Caching), #17 (Observability), #18 (Resilience)
- Coordinates with: #11 (Configuration), #19 (Async Resources)
- Depends on: PR #23 (recommended base branch)

---

**Status**: 🟡 Ready to Start  
**Priority**: Medium (can proceed in parallel with other work)  
**Estimated Effort**: 3-4 weeks (phased implementation)  
**Skills Required**: Python, PyTorch, CUDA, Async Programming
